pinot-query-runtime/src/main/java/org/apache/pinot/query/mailbox/ReceivingMailbox.java

```diff
@@ -18,55 +18,118 @@
+          LOGGER.debug("Mailbox: {} is already cancelled or errored out, ignoring the late block", _id);
+          _blocks.clear();
+          return false;
+        }
+      } else {
+        LOGGER.debug("Failed to offer block into mailbox: {} within: {}ms", _id, timeoutMs);
+        setErrorBlock(TransferableBlockUtils.getErrorTransferableBlock(
+            new TimeoutException("Timed out while waiting for receive operator to consume data from mailbox: " + _id)));
+        return false;
+      }
+    } catch (InterruptedException e) {
+      LOGGER.error("Interrupted while offering block into mailbox: {}", _id);
+      setErrorBlock(TransferableBlockUtils.getErrorTransferableBlock(e));
+      return false;
+    }
+  }
 
   /**
-   * A ReceivingMailbox is considered initialized when it has a reference to the underlying channel used for receiving
-   * the data. The underlying channel may be a gRPC stream, in-memory queue, etc. Once a receiving mailbox is
-   * initialized, it has the ability to close the underlying channel via the {@link #cancel()} method.
+   * Sets an error block into the mailbox. No more blocks are accepted after calling this method.
    */
-  boolean isInitialized();
+  public void setErrorBlock(TransferableBlock errorBlock) {
+    if (_errorBlock.compareAndSet(null, errorBlock)) {
+      _blocks.clear();
+      _receiveMailCallback.accept(_id);
+    }
+  }
 
   /**
-   * A ReceivingMailbox is considered closed if it has sent all the data to the receiver and doesn't have any more data
-   * to send.
+   * Returns the first block from the mailbox, or {@code null} if there is no block received yet. Error block is
+   * returned if exists.
    */
-  boolean isClosed();
+  @Nullable
+  public TransferableBlock poll() {
+    TransferableBlock errorBlock = _errorBlock.get();
+    return errorBlock != null ? errorBlock : _blocks.poll();
+  }
 
   /**
-   * A ReceivingMailbox may hold a reference to the underlying channel. Usually the channel would be automatically
-   * closed once all the data has been received by the receiver, and in such cases {@link #isClosed()} returns true.
-   * However in failure scenarios the underlying channel may not be released, and the receiver can use this method to
-   * ensure the same.
-   *
-   * This API should ensure that the underlying channel is "released" if it hasn't been already. If the channel has
-   * already been released, the API shouldn't throw and instead return gracefully.
-   *
-   * <p>
-   *   This method may be called multiple times, so implementations should ensure this is idempotent.
-   * </p>
+   * Cancels the mailbox. No more blocks are accepted after calling this method. Should only be called by the receive
+   * operator to clean up the remaining blocks.
    */
-  void cancel();
+  public void cancel() {
+    LOGGER.debug("Cancelling mailbox: {}", _id);
+    if (_errorBlock.compareAndSet(null, CANCELLED_ERROR_BLOCK)) {
+      _blocks.clear();
+    }
+  }
```

Comment:
should accept upstream throwable here
```suggestion
  public void cancel(Throwable t) {
    LOGGER.debug("Cancelling mailbox: {}", _id, t);
    if (_errorBlock.compareAndSet(null, TransferableBlockUtils.getErrorTransferableBlock(t))) {
      _blocks.clear();
    }
  }
```